templates/_worker.js/matcher.ts

```diff
@@ -0,0 +1,414 @@
+			this.ctx
+		);
+
+		if (resp.status >= 400) {
+			// The middleware function errored. Set the status and bail out.
+			this.status = resp.status;
+			return false;
+		}
+
+		this.processMiddlewareResp(resp);
+		return true;
+	}
+
+	/**
+	 * Resets the response status and headers if the route should override them.
+	 *
+	 * @param route Build output config source route.
+	 */
+	private applyRouteOverrides(route: VercelSource): void {
+		if (!route.override) return;
+
+		this.status = undefined;
+		this.headers.normal = new Headers();
+		this.headers.important = new Headers();
+	}
+
+	/**
+	 * Applies the route's headers for the response object.
+	 *
+	 * @param route Build output config source route.
+	 * @param srcMatch Matches from the PCRE matcher.
+	 * @param captureGroupKeys Named capture group keys from the PCRE matcher.
+	 */
+	private applyRouteHeaders(
+		route: VercelSource,
+		srcMatch: RegExpMatchArray,
+		captureGroupKeys: string[]
+	): void {
+		if (!route.headers) return;
+
+		applyHeaders(route.headers, this.headers.normal, {
+			match: srcMatch,
+			captureGroupKeys,
+		});
+
+		if (route.important) {
+			applyHeaders(route.headers, this.headers.important, {
+				match: srcMatch,
+				captureGroupKeys,
+			});
+		}
+	}
+
+	/**
+	 * Applies the route's status code for the response object.
+	 *
+	 * @param route Build output config source route.
+	 */
+	private applyRouteStatus(route: VercelSource): void {
+		if (!route.status) return;
+
+		this.status = route.status;
+	}
+
+	/**
+	 * Applies the route's destination for the matching the path to the Vercel build output.
+	 *
+	 * @param route Build output config source route.
+	 * @param srcMatch Matches from the PCRE matcher.
+	 * @param captureGroupKeys Named capture group keys from the PCRE matcher.
+	 * @returns The previous path for the route before applying the destination.
+	 */
+	private applyRouteDest(
+		route: VercelSource,
+		srcMatch: RegExpMatchArray,
+		captureGroupKeys: string[]
+	): string {
+		if (!route.dest) return this.path;
+
+		const prevPath = this.path;
+
+		this.path = applyPCREMatches(route.dest, srcMatch, captureGroupKeys);
+
+		// NOTE: Special handling for `/index` RSC routes. Sometimes the Vercel build output config
+		// has a record to rewrite `^/` to `/index.rsc`, however, this will hit requests to pages
+		// that aren't `/`. In this case, we should check that the previous path is `/`.
+		if (/\/index\.rsc$/i.test(this.path) && !/\/(?:index)?$/i.test(prevPath)) {
+			this.path = prevPath;
+		}
+
+		// NOTE: Special handling for `.rsc` requests. If the Vercel CLI failed to generate an RSC
+		// version of the page and the build output config has a record mapping the request to the
+		// RSC variant, we should strip the `.rsc` extension from the path.
+		if (/\.rsc$/i.test(this.path) && !(this.path in this.output)) {
+			this.path = this.path.replace(/\.rsc/i, '');
+		}
+
+		// If not an external URL, merge search params for later use.
```

Comment:
If it's an external URL, we don't apply the search parameters before redirecting. We only apply search parameters to rewritten URLs so that they can be consumed by the page/function when it is run.

Perhaps this behaviour should be changed so that we always apply search parameters, regardless of whether it's a full URL or not. What do you think? There wasn't a specific reason I didn't do that, so maybe I should?

Reply:
yeah sounds like that could be simpler (and if it turns out problematic we can always add this external check back in)

I would go with that if you're ok with it, just for the sake of keeping things (slightly) simpler and not having extra logic which might not be needed.

Anyways If you want I am also ok with the current implementation, I was asking `why` just because honestly I wasn't sure as to why the check is here, I wasn't saying that it is wrong or anything. But if you do decide to keep the check please expand the comment to explain also the reason as to why we do this :slightly_smiling_face: :pray: